prime-router/src/main/kotlin/azure/ActionHistory.kt

```diff
@@ -0,0 +1,336 @@
+import com.microsoft.azure.functions.ExecutionContext
+import com.microsoft.azure.functions.HttpRequestMessage
+import com.microsoft.azure.functions.HttpResponseMessage
+import gov.cdc.prime.router.ClientSource
+import gov.cdc.prime.router.OrganizationService
+import gov.cdc.prime.router.Report
+import gov.cdc.prime.router.azure.db.Tables.ACTION
+import gov.cdc.prime.router.azure.db.Tables.REPORT_FILE
+import gov.cdc.prime.router.azure.db.Tables.REPORT_LINEAGE
+import gov.cdc.prime.router.azure.db.enums.TaskAction
+import gov.cdc.prime.router.azure.db.tables.pojos.Action
+import gov.cdc.prime.router.azure.db.tables.pojos.ReportFile
+import gov.cdc.prime.router.azure.db.tables.pojos.ReportLineage
+import org.jooq.Configuration
+import org.jooq.impl.DSL
+import java.io.ByteArrayOutputStream
+import java.lang.IllegalArgumentException
+import java.util.UUID
+
+/**
+ * This is a container class that holds information to be stored, about a single action,
+ * as well as the reports that went into that Action, and were created by that Action.
+ *
+ * The idea is that, as an action progresses, call various track*(...) methods here to add additional information to
+ * this container, in-memory only.
+ *
+ * Then when the action is done, call saveToDb(...) to plunk all the tracked information into the database.
+ *
+ */
+class ActionHistory {
+
+    private var context: ExecutionContext?
+
+    /**
+     * Throughtout, using generated mutable jooq POJOs purely as a convenience to storage information
+     *
+     */
+
+    var action = Action()
+
+    /*
+     * Reports that are inputs to this action, from previous steps (already in database)
+     */
+    var reportsIn = mutableMapOf<UUID, ReportFile>()
+
+    /*
+     * Reports that are inputs to this action, from external source (not already in database)
+     * Note that this should be able to handle multiple submitted reports in one action.
+     */
+    var reportsReceived = mutableMapOf<UUID, ReportFile>()
+
+    /*
+     * New reports generated by this action.  (also not yet in database)
+     */
+    var reportsOut = mutableMapOf<UUID, ReportFile>()
+
+    /**
+     *
+     * Collection of all the parent-child report relationships created by this action.
+     *
+     * Note:  There is a strong OO argument that this list should be broken out into each individual child Report.kt.
+     * (That is, every report should know its own parents!)
+     * However, its here because there are Functions that do not create Report.kt objects.  For example, Send.
+     * In addition, in-memory, reports get copied many times, with lots of parent-child relationships
+     * that are error-prone to track.  Hiding the lineage data here helps ensure correctness and hide complexity.
+     *
+     * todo Note that this does not work for command line.   Is that a problem?
+     * todo this is redundant with `Report.sources`.   Merge these together; eliminate one of them.
+     *
+     */
```

Comment:
Each time some new morsel of historical information is created, call one of the "track" functions to store it here.

Its like the Presidential Records Act, but in Kotlin.